staker/fast_confirm.go

```diff
@@ -0,0 +1,223 @@
+		big.NewInt(0),
+		big.NewInt(0),
+		common.Address{},
+		common.Address{},
+		nonce,
+	)
+	if err != nil {
+		return err
+	}
+	if !f.wallet.CanBatchTxs() {
+		err = f.flushTransactions(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	auth, err := f.builder.Auth(ctx)
+	if err != nil {
+		return err
+	}
+	_, err = f.safe.ApproveHash(auth, safeTxHash)
+	if err != nil {
+		return err
+	}
+	if !f.wallet.CanBatchTxs() {
+		err = f.flushTransactions(ctx)
+		if err != nil {
+			return err
+		}
+	}
+	return f.checkApprovedHashAndExecTransaction(ctx, fastConfirmCallData, safeTxHash)
+}
+
+func (f *FastConfirmSafe) flushTransactions(ctx context.Context) error {
+	arbTx, err := f.wallet.ExecuteTransactions(ctx, f.builder, f.gasRefunder)
+	if err != nil {
+		return err
+	}
+	if arbTx != nil {
+		_, err = f.l1Reader.WaitForTxApproval(ctx, arbTx)
+		if err == nil {
+			log.Info("successfully executed staker transaction", "hash", arbTx.Hash())
+		} else {
+			return fmt.Errorf("error waiting for tx receipt: %w", err)
+		}
+	}
+	f.builder.ClearTransactions()
+	return nil
+}
+
+func (f *FastConfirmSafe) createFastConfirmCalldata(
+	blockHash common.Hash, sendRoot common.Hash,
+) ([]byte, error) {
+	calldata, err := f.fastConfirmNextNodeMethod.Inputs.Pack(
+		blockHash,
+		sendRoot,
+	)
+	if err != nil {
+		return nil, err
+	}
+	fullCalldata := append([]byte{}, f.fastConfirmNextNodeMethod.ID...)
+	fullCalldata = append(fullCalldata, calldata...)
+	return fullCalldata, nil
+}
+
+func (f *FastConfirmSafe) checkApprovedHashAndExecTransaction(ctx context.Context, fastConfirmCallData []byte, safeTxHash [32]byte) error {
+	var signatures []byte
+	approvedHashCount := uint64(0)
+	for _, owner := range f.owners {
+		if f.wallet.Address() == nil {
+			return errors.New("wallet address is nil")
+		}
+		var approved *big.Int
+		// No need check if wallet has approved the hash,
+		// since checkApprovedHashAndExecTransaction is called only after wallet has approved the hash.
+		if *f.wallet.Address() == owner {
+			approved = common.Big1
+		} else {
+			var err error
+			approved, err = f.safe.ApprovedHashes(&bind.CallOpts{Context: ctx}, owner, safeTxHash)
+			if err != nil {
+				return err
+			}
+		}
+
+		// If the owner has approved the hash, we add the signature to the transaction.
+		// We add the signature in the format r, s, v.
+		// We set v to 1, as it is the only possible value for a approved hash.
+		// We set r to the owner's address.
+		// We set s to the empty hash.
+		// Refer to the Safe contract for more information.
+		if approved.Cmp(common.Big1) == 0 {
+			approvedHashCount++
+			v := uint8(1)
+			r := common.BytesToHash(owner.Bytes())
+			s := common.Hash{}
+			signatures = append(signatures, r.Bytes()...)
+			signatures = append(signatures, s.Bytes()...)
+			signatures = append(signatures, v)
+		}
+	}
+	if approvedHashCount >= f.threshold {
+		auth, err := f.builder.Auth(ctx)
+		if err != nil {
+			return err
+		}
+		_, err = f.safe.ExecTransaction(
```

Comment:
I just realized one other issue here. What if the final two committee members make a tx in the same block, so none of them call ExecTransaction, because they're all looking at the prior block's state where there wasn't quorum yet? I'm not sure exactly how to resolve this. The latency optimal way would be to have a smart contract that determines if ExecTransaction can be called, but I think a more realistic option would be for the node to circle back after its tx is confirmed and try to ExecTransaction if it hasn't been done yet.

Reply:
Makes sense, updated according to second approach.